contributors/design-proposals/auth/pod-restriction.md

```diff
@@ -0,0 +1,477 @@
+PodRestriction controller. The registration should include a path (auto-discovery?) to extract the
+PodTemplate from the CRD.
+
+### API Design
+
+#### Monolith vs. Modular
+
+Our current approach to policies has been "modular", i.e. a different controller and API for each
+"area" of features. This has led to some inconsistencies and gaps in our policies, and makes adding
+new restrictions needlessly complex. I propose we take a "monolithic" approach, with a single policy
+(PodRestriction) for all Pod fields.
+
+A monolithic restriction policy (i.e. single resource type) also allows the user to express more
+complex inter-relationships, such as:
+
+-   Pods using a certain volume type must run with reduced privileges
+
+-   Pods can only run as privileged if they run with a specific AntiAffinity
+
+-   Pods labeled as staging must not run on production nodes.
+
+#### No Defaulting
+
+There are 2 problems with mutating Pod fields: 1) it becomes hard to compose policies (what happens
+when the defaults conflict?) and 2) you need a policy ordering mechanism for when a Pod matches
+multiple policies. Instead, PodRestriction opts out of mutating defaults entirely, and leaves that
+for a different component to solve ([PodPreset?](../service-catalog/pod-preset.md)).
+
+## API Sketch
+
+I don't want to hash out all the details of the API (such as which fields are constrained, and what
+the naming is) in this proposal, but there are a couple high level design goals and takeaways:
+
+1.  The PodRestriction API layout mostly mirrors that of the Pod object. This makes the API more
+    predictable, and easier to manage (i.e. I want to restrict Pod field X, I know how to do that in
+    the PodRestriction).
+
+2.  Restrictions are self-contained. There isn't complex interdependent logic here. If you want more
+    than basic whitelist / blacklist / pattern matching, then implement a new policy or controller
+    for that.
+
+3.  There are a lot of fields in the PodSpec, and thus this API gets large (see [Monolith
+    vs. Modular](#monolith-vs-modular)). We don't need to build it all at once, but I think over
+    time it will eventually grow into the full version.
+
+4.  Fields are unrestricted by default. This is important due to the size of the API. If I want to
+    restrict 1 field, I shouldn't need to write 500 lines to open up all the other fields.
+
+```go
+// This API sketch is partial, and only intended to indicate the the flavor of
+// the API, not to completely specify it.
+
+type PodRestriction struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+    Binding *PodRestrictionBinding // Default to cluster-wide
+
+    Spec PodRestrictionSpec
+}
+
+type PodRestrictionBinding struct {
+    Mode              BindingMode  // Default = Drop
+
+    Namespaces        []string
+    NamespaceSelector *meta.LabelSelector
+}
+
+type BindingMode string
+
+const (
+    BindingModeAccept BindingMode = "Accept"
+    BindingModeDrop   BindingMode = "Drop"
```

Comment:
I find the term "Drop" confusing. It's Deny, the opposite of Accept? (As opposed to, e.g., elide.)